hibernate-core/src/main/java/org/hibernate/dialect/function/CastFunction.java

```diff
@@ -77,6 +77,9 @@ public void render(
 			renderCastArrayToString( sqlAppender, arguments.get( 0 ), dialect, walker );
 		}
 		else {
+			if ( source instanceof QueryLiteral<?> literal && literal.getLiteralValue() == null ) {
+				literal.setInCast(true);
```

Comment:
You're making `QueryLiteral` mutable and change its state as part of rendering here which is a bit dirty IMO. Can you explain why you need this?`